### PR TITLE
Performance improvements for Collision Pick API

### PR DESCRIPTION
**Pull request by sabrina-shanman**
This is a performance improvement for https://github.com/highfidelity/hifi/pull/13714. The most notable change adds the passing of entity/avatar flags to bullet which allows bullet to eliminate search areas earlier on. Some data copying is also reduced.

## TEST PLAN
- Use this manual test for the following instructions: https://github.com/sabrina-shanman/hifi_tests/tree/safe-teleport-target/tests/engine/interaction/pick/collision/capsule_on_server
    - Perform the test in master, keeping track of the locations used for the next test. As master has already been tested, assume the capsule pick behaves correctly and just record the pickManager times.
    - Perform the test in this PR, using measurements in the same locations as were used in master.
    - Compare recorded times. Assume differences of less than 15% times the "original value" for the pickManager time are negligible.  (So, for example, if the base pickManager time is 1.7 ms, a difference of 0.2ms is considered negligible) Expect an overall performance gain (lower times in ms).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12146/HighFidelity-Beta-PR13833-67ff057.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12146/HighFidelity-Beta-Interface-PR13833-67ff057.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12147/HighFidelity-Beta-PR13833-130cb70.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12147/HighFidelity-Beta-PR13833-130cb70.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12147/HighFidelity-Beta-Interface-PR13833-130cb70.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12147/HighFidelity-Beta-PR13833-130cb70.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12147/HighFidelity-Beta-Interface-PR13833-130cb70.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Stopped Testing pending further information.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12155/HighFidelity-Beta-PR13833-b655727.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12155/HighFidelity-Beta-PR13833-b655727.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12155/HighFidelity-Beta-Interface-PR13833-b655727.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12155/HighFidelity-Beta-PR13833-b655727.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12155/HighFidelity-Beta-Interface-PR13833-b655727.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
@sabrina-shanman  Still getting error when running script.
[7/17 9:47:16] [testStory.js] ERROR - [UncaughtException evaluate] Error: error retrieving script (NotFound) in https://raw.githubusercontent.com/highfidelity/hifi_tests/master/tests/utils/branchUtils.js:52
[Backtrace]
    <anonymous>(executionPath = 'https://github.com/sabrina-shanman/hifi_tests/raw/safe-teleport-target/tests/engine/interaction/pick/collision/capsule_on_server/') at https://raw.githubusercontent.com/highfidelity/hifi_tests/master/tests/utils/branchUtils.js:52
    <global>() at https://github.com/sabrina-shanman/hifi_tests/raw/safe-teleport-target/tests/engine/interaction/pick/collision/capsule_on_server/testStory.js:3

In addition to this, the script linked in the test story has an invalid URL.

**Comment by sabrina-shanman** [comment]
I fixed the broken link. Still at a loss of why the script isn't loading properly.

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Times are lower in this PR resulting in better performance.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12172/HighFidelity-Beta-PR13833-6081d63.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12172/HighFidelity-Beta-PR13833-6081d63.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12172/HighFidelity-Beta-Interface-PR13833-6081d63.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12172/HighFidelity-Beta-PR13833-6081d63.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12172/HighFidelity-Beta-Interface-PR13833-6081d63.exe

**Comment by sabrina-shanman** [comment]
Andrew pointed out that boolean conversion of std::vector::size() didn't follow the coding standards, so I addressed that.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12176/HighFidelity-Beta-PR13833-5194089.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12176/HighFidelity-Beta-PR13833-5194089.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12176/HighFidelity-Beta-Interface-PR13833-5194089.dmg

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12176/HighFidelity-Beta-Interface-PR13833-5194089.exe

File: interface/src/raypick/CollisionPick.h [hunk 1]
```diff
@@ -41,31 +45,15 @@ class CollisionPickResult : public PickResult {
 
     LoadState loadState { LOAD_STATE_UNKNOWN };
     bool intersects { false };
-    std::vector<ContactTestResult> entityIntersections;
-    std::vector<ContactTestResult> avatarIntersections;
+    std::shared_ptr<std::vector<ContactTestResult>> entityIntersections { std::make_shared<std::vector<ContactTestResult>>() };
+    std::shared_ptr<std::vector<ContactTestResult>> avatarIntersections { std::make_shared<std::vector<ContactTestResult>>() };
```
Comment (SamGondelman) [inline interface/src/raypick/CollisionPick.h]: These don't need to be `std::shared_ptr<std::vector>`

**Comment by conklin94122** [comment]
Withdrawing QA Approved until Sam's changes are made. If those changes require zero retesting let me know I'll put the label back. Otherwise please label as QA Ready after the changes have been made.

**Comment by qaAdamSmith82** [comment]
Build this please

File: interface/src/raypick/CollisionPick.cpp [hunk 2]
```diff
@@ -308,20 +332,24 @@ CollisionRegion CollisionPick::getMathematicalPick() const {
...
+    }
+
+    int n = (int)intersections.size();
+    for (int i = 0; i < n; i++) {
+        auto& intersection = intersections[i];
         const QUuid& id = intersection.foundID;
-        if (!ignoreItems.contains(id) && (!isWhitelist || includeItems.contains(id))) {
-            filteredIntersections.push_back(intersection);
+        if (ignoreItems.contains(id) || (isWhitelist && !includeItems.contains(id))) {
+            intersections[i] = intersections[--n];
```
Comment (SamGondelman) [inline interface/src/raypick/CollisionPick.cpp]: What if intersections[--n] would be filtered out?

**Comment by conklin94122** [comment]
Changes requested, pulling QA Ready

File: interface/src/raypick/CollisionPick.cpp (hunk 2 above)
Comment (sabrina-shanman) [reply interface/src/raypick/CollisionPick.cpp]: Better fix that, then.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12197/HighFidelity-Beta-PR13833-813290a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12197/HighFidelity-Beta-Interface-PR13833-813290a.exe

**Comment by sabrina-shanman** [comment]
Fixed a build error with Mac/Linux/Android

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12203/HighFidelity-Beta-PR13833-9c36519.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12203/HighFidelity-Beta-PR13833-9c36519.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12203/HighFidelity-Beta-Interface-PR13833-9c36519.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12203/HighFidelity-Beta-PR13833-9c36519.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12203/HighFidelity-Beta-Interface-PR13833-9c36519.exe

**Comment by qaAdamSmith82** [comment]
Currently Testing

**Comment by qaAdamSmith82** [comment]
Pass @conklin94122 
Original Value 2.7ms
Non-Colliding Capsule 3.7ms, 3.8ms, 3.8ms
Static Mesh Colliding Capsule 4.2ms
Value After Cleanup 2.8ms

**Pull request merged**